_posts/2020-12-21--kinesis-streams-part-2.md

```diff
@@ -0,0 +1,591 @@
+
+The _GetRecords API_ comes with some important [limits](https://docs.aws.amazon.com/streams/latest/dev/service-sizes-and-limits.html) we haven’t discussed yet. First of all, there is a limit on **how much data** can a single _GetRecords_ call retrieve from a shard (either 10 000 records or 10 MB, whichever comes first). The good thing is, in the case of Lambda, the _ESM_ automatically makes sure this limit is never reached (remember, you can not have a batch bigger than 10 000 records?).
+
+The second, and a somewhat more noticeable limit, is that there can be **at most 5 _GetRecords_ calls** made per second in each shard. Once again, the good thing is, this is already taken care of by the _ESM_ that only polls the shards once per second. However, this limit also means that you can have at most **5 different consumers** using the shared throughput before they start to get **throttled**.
+
+
+![shared_throughput](/img/kinesis/shared_throughput.png){: .img.centered }
+
+
+When throttling (or other intermittent error) happens with _Lambda service_, the _ESM_ steps in and handles this behind the scenes with some built-in retries and exponential backoffs. Note, that the maximum record age you set for the _ESM_ also applies in this case and older records will not be sent to your Lambda function. The records will likely be processed eventually, but the retries will slow down your stream consumption and may lead to records expiring from the stream over time.
+
+All this means, that while technically there’s no limit on how many Lambda functions you can attach to a single stream while using the shared throughput, in reality
+- the 5 _GetRecords_ call limit means you can have at most 5 consumers before they start to get **throttled**, and
+- the 2 MB throughput limit means you might start **getting behind your stream** if u have more than 2 consumers.
+
+The 5 request limit also means that the records can be read from the stream roughly once every **200 milliseconds** (1 000 ms / 5 calls). This is why the average propagation latency for a shared throughput consumer is estimated at roughly 200 ms. This is also the reason, why adding more consumers means **increasing the propagation latency** to up to 1 second.
+
+This all explains why, if you need to have more than 2 to 3 separate consumers reading from your stream, or you need to have really low, near real-time read latencies, enhanced fan-out is the way to go.
```

Comment:
I'd write: "This all explains why enhanced fan-out is the way to go if you need to have more than 2 to 3 separate consumers reading from your stream, or you need to have really low, near real-time read latencies."